File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.java
class Pair{

    int sum;
    int i;
    int j;
    Pair(int sum, int i, int j){
        this.sum = sum;
        this.i = i;
        this.j = j;
    }
}

class Solution {
    public List<List<Integer>> kSmallestPairs(int[] a, int[] b, int k) {
        
        List<List<Integer>> result = new ArrayList<>();
        int n = a.length;
        int m = b.length;
        
        PriorityQueue<Pair> pq = new PriorityQueue<>((p1, p2) -> p2.sum - p1.sum);
        
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                int sum = a[i] + b[j];
                
                if(pq.size()<k){
                    pq.offer(new Pair(sum, i, j));
                }
                else if(pq.peek().sum>sum){
                    pq.poll();
                    pq.offer(new Pair(sum, i, j));
                }
                else{
                    break;
                }
            }
        }
        
        while(!pq.isEmpty()){
            Pair p = pq.poll();
            List<Integer> list = new ArrayList<>();
            list.add(a[p.i]);
            list.add(b[p.j]);
            result.add(list);
        }
        
        return result;
    }
}

File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.cpp
class Solution {
public:
    vector<vector<int>> brute(vector<int>& a, vector<int>& b, int k){
        
        int n = a.size();
        int m = b.size();
        
        vector<vector<int>> sums;
        
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                vector<int> temp = {a[i] + b[j], a[i], b[j]};
                sums.push_back(temp);
            }
        }
        
        sort(begin(sums), end(sums));
        
        vector<vector<int>> result;
        
        for(int i=0; i<min(k, (int)sums.size()); i++){
            vector<int> temp = {sums[i][1], sums[i][2]};
            result.push_back(temp);
        }
        return result;
    }
    
    vector<vector<int>> kSmallestPairs(vector<int>& a, vector<int>& b, int k) {
        
        int n = a.size();
        int m = b.size();
        
        priority_queue<pair<int, pair<int, int>> , vector<pair<int, pair<int, int>>>, greater<pair<int, pair<int, int>>>> pq;
        
        set<pair<int, int>> visited;
        
        pq.push({a[0]+b[0], {0, 0}});
        
        vector<vector<int>> result;
        
        while(k-- and !pq.empty()){
            
            auto top = pq.top(); pq.pop();
            
            int i = top.second.first;
            int j = top.second.second;
            result.push_back({a[i], b[j]});
            
//             cout << i <<" " << j << endl;
//             cout << i+1 << " " << j << " / " << i << " " << j+1 << endl; 
            
            if(i+1<n and visited.find({i+1, j}) == visited.end()){
                pq.push({a[i+1]+b[j], {i+1, j}});
                visited.insert({i+1, j});
            }
            
            if(j+1<m and visited.find({i, j+1}) == visited.end()){
                pq.push({a[i]+b[j+1], {i, j+1}});
                visited.insert({i, j+1});
            }
        }
        return result;
    }
};